src/extras/transform.js

```diff
@@ -1,32 +1,75 @@
+import { detectFormat} from '../utils/formats';
+
 /*
  * Support for a "transform" loader interface
  */
-(function () {
-  const systemJSPrototype = System.constructor.prototype;
-
-  const instantiate = systemJSPrototype.instantiate;
-  systemJSPrototype.instantiate = function (url, parent) {
-    if (url.slice(-5) === '.wasm')
-      return instantiate.call(this, url, parent);
-
-    const loader = this;
-    return fetch(url, { credentials: 'same-origin' })
-    .then(function (res) {
-      if (!res.ok)
-        throw new Error('Fetch error: ' + res.status + ' ' + res.statusText + (parent ? ' loading from ' + parent : ''));
-      return res.text();
+const systemJSPrototype = System.constructor.prototype;
+
+const resolve = systemJSPrototype.resolve;
+systemJSPrototype.resolve = function(url, parent) {
+  return resolve.call(this, url, parent);
+};
+
+const instantiate = systemJSPrototype.instantiate;
+systemJSPrototype.instantiate = function (url, parent) {
+  if (url.slice(-5) === '.wasm') {
+    return instantiate.call(this, url, parent);
+  }
+
+  const context = { parent };
+
+  return this.fetch(context, url)
+    .catch(({ code, message }) => {
+      throw new Error(`Fetch error: ${code} ${message}${parent ? ` loading from ${parent}` : ''}`);
     })
-    .then(function (source) {
-      return loader.transform.call(this, url, source);
+    .then(source => this.detectFormat(context, url, source))
+    .then(source => {
+      if (this.transform.length === 3) {
+        return this.transform(context, url, source);
+      }
+      return this.transform(url, source);
     })
-    .then(function (source) {
-      (0, eval)(source + '\n//# sourceURL=' + url);
-      return loader.getRegister();
-    });
-  };
+    .then(source => this.evaluate(context, url, source))
+    .then(registration => registration || this.getRegister());
+};
+
 
-  // Hookable transform function!
-  systemJSPrototype.transform = function (_id, source) {
-    return source;
+// Hookable fetch function
+systemJSPrototype.fetch = function(context, url) {
+  const options = {
+    credentials: 'same-origin',
   };
-})();
\ No newline at end of file
+
+  return fetch(url, options).then(function (res) {
+    if (res.ok) {
+      return res.text();
+    }
+
+    throw {
+      url,
+      code: res.status,
+      message: res.statusText,
+    };
+  }).catch(function(err) {
+    throw err;
+  });
+};
+
+
+// Hookable file format detection
+systemJSPrototype.detectFormat = function(context, url, source) {
+  context.format = detectFormat(url, source);
+  return source;
+};
+
+
+// Hookable transform function!
+systemJSPrototype.transform = function(context, url, source) {
```

Comment:
I like the idea of a context argument, but the way it is done here is not backwards compatible. Also it should be its own PR.